src/sentry/static/sentry/app/views/organizationDiscover/discover.jsx

```diff
@@ -116,53 +142,162 @@ export default class OrganizationDiscover extends React.Component {
 
     clearIndicators();
 
-    resultManager
-      .fetchAll()
-      .then(data => {
-        this.setState({data, isFetchingQuery: false});
+    resultManager.fetchAll().then(data => {
+      const shouldRedirect = !savedQuery;
 
+      if (shouldRedirect) {
         browserHistory.push({
           pathname: `/organizations/${organization.slug}/discover/`,
           search: getQueryStringFromQuery(queryBuilder.getInternal()),
         });
+      }
+
+      this.setState({data, isFetchingQuery: false});
+    });
+  };
+
+  toggleSidebar = view => {
+    if (view !== this.state.view) {
+      this.setState({view});
+      browserHistory.replace({
+        pathname: `/organizations/${this.props.organization.slug}/discover/`,
+        query: {...this.props.location.query, view},
+      });
+    }
+  };
+
+  toggleEditMode = () => {
+    this.setState(state => {
+      const isEditMode = !state.isEditingSavedQuery;
+      return {
+        isEditingSavedQuery: isEditMode,
+        savedQueryName: isEditMode ? this.props.savedQuery.name : null,
+      };
+    });
+  };
+
+  reset = () => {
+    const {savedQuery, queryBuilder} = this.props;
+    if (savedQuery) {
+      queryBuilder.reset(savedQuery);
+      this.setState({
+        isEditingSavedQuery: false,
+      });
+    } else {
+      browserHistory.push({
+        pathname: `/organizations/${this.props.organization.slug}/discover/`,
```

Comment:
small nit, can we destructure `organization` here since we're doing it for other props?